js/language_server/Main.ml

```diff
@@ -0,0 +1,105 @@
+(* Austin Theriault
+ *
+ * Copyright (C) 2019-2023 Semgrep, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(* Commentary *)
+(*  *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+
+open Js_of_ocaml
+open Semgrep_js_shared
+
+let read_line_ref = ref (fun () -> Lwt.return None)
+let read_exactly_ref = ref (fun _ -> Lwt.return None)
+
+module Io = RPC_server.MakeLSIO (struct
+  type input = unit
+  type output = unit
+
+  let read_line _ = !read_line_ref ()
+  let stdin = ()
+  let stdout = ()
+  let flush _ = Lwt.return ()
+  let atomic f oc = f oc
+
+  let write _ str =
+    (* nosem *)
+    print_string str;
+    Lwt.return ()
```

Comment:
Confirm? Do you want the printing to occur with the promise?

Reply:
Yea, it's a consequence of how the LSP library does things